### DISCUSSION TITLE
USB Soundcard emulation

### DISCUSSION THREAD

**Discussion by goopaga** [kamilsss655/ESPRI#33] [Ideas]
Emulation of an USB soundcard would open a wide range of new possibilities such as enabling Quanshengdock, APRS, etc.
Virtually providing also the features of AIOC (All In One Cable) and many more thanks to the great work done so fare and what is about to come.

Of course the “regular” ESP32 on Lolin Lite cannot provide such feature as the USB connector is exposing just the serial COM via the TTL-to-USB adapter and not a fully compliant USB.

AFAIK the ESP32-S3 has a “fully fledged” USB interface and a mixed mode (audio in/out + data) could be possible.

I don’t know if this is feasible and from a quick Google search the actual showstoppers could be:
- lack of DAC on ESP32-S3 (an external DAC could be required ?)
- according to some issues reported on the Espressif account on GitHub the TynyUSB component has some missing parts being worked at the moment for this featur (see https://github.com/espressif/esp-idf/issues/12774).

What do you think ?

**Comment by kamilsss655** [comment]
Not a high priority now, but could be added as 'nice to have' in later stages of the project.

**Comment by kamilsss655** [comment]
Same with USB to SERIAL functionality to use with CHIRP.

**Reply by goopaga** [reply]
For this purpose, would it be possible to set the ESPRI (via webserver command) in a sort of pass through mode so that the COM port is “detached” from the terminal monitor and can be used to read/write the radio memory (CHIRP) ? Regular ESPRI operations could be restored with another command via webserver or on reboot.

I don’t want to appear demanding as I’m not able to help with coding, however this section should be the right place to drop some ideas for future developments and I hope it’s appreciated.
I admire your magnificent work with this project.

**Reply by kamilsss655** [reply]
Yea it might be as simple as copying all the UART data between the two UART ports.

**Comment by goopaga** [comment]
Maybe the DAC is not necessary. A GPIO can be used to output PDM (Pulse Density Modulation) and then filtered with an RC to “reconstruct” the signal.
For future reference :
https://atomic14.substack.com/p/esp32-s3-no-dac

**Reply by kamilsss655** [reply]
Yup, I am not using DAC in ESPRI. I used PDM before wav beacon, and now using PWM audio.
Although I might come back to PDM, as it seems to produce better quality audio, but need to confirm that.

**Reply by kamilsss655** [reply]
BTW the sound recorder (WIP) seems to work really well. I was expecting noise, but audio is really good.
